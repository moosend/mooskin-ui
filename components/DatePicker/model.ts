import { DatePickerProps, KeyboardDatePickerProps } from '@material-ui/pickers';
import { IInputComponentProps } from '../Input/model';

export type PickerType = 'date' | 'date-keyboard';

export interface IDatePickerCommonProps {
	inputComponentProps?: IInputComponentProps;
	pickerType?: PickerType;
}

export interface IDatePickerComponentProps extends IDatePickerCommonProps, DatePickerProps {}
export interface IDatePickerKeyboardComponentProps extends IDatePickerCommonProps, KeyboardDatePickerProps {}
