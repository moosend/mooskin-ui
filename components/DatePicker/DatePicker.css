@import '../_utils/globals/screens.css';
@import '../_utils/globals/variables.css';
@import url('https://code.ionicframework.com/ionicons/2.0.1/css/ionicons.min.css');

.datePicker{
    display: flex;
    position: relative;
    flex: 1;
}

.wrapper{
    display: flex;
    flex: 1;
    flex-direction: column;
}

.label{
    display: flex;
    flex-direction: column;
    justify-content: center;
    color: #6b6b6b;
     /* font-size: 13px;  */
     font-size: .92857142857rem; 
    cursor: pointer;
    flex-basis: 150px;
}

.dateInput{
    outline: none;
    padding: 15px;
    font-family: Arial,sans-serif;
    border: 0;
    color: var(--inputFontColor);
    /* font-size: 16px; */
    font-size: 1rem; 
    border-radius: 8px;
    background: #efefef;
    flex: 1;
    cursor: pointer;
}

.disabled{
    cursor: not-allowed;
    color: #bfbfbf;
    background: #f6f6f6;
}
 
.calendar{
    position: absolute;
    top: 50px;
    z-index: 99;
    box-shadow: 1px 10px 20px rgba(0,0,0,.2);
}

.cover{
    position: fixed;
    z-index: -1;
    top: 0px;
    right: 0px;
    bottom: 0px;
    left: 0px;
}

:global(.datepicker-component .m-input-moment){
    background-color: #ffffff;
    user-select: none;
    height: 355px;
}

:global(.datepicker-component .m-input-moment .tabs){
    margin: 0;
}

:global(.datepicker-component .m-input-moment .options button.is-active){
    background: #5ccdde;
    color: #fff;
}

:global(.datepicker-component .m-input-moment .options button){
    border: 1px solid #5ccdde;
    color: #5ccdde;
    margin: 0px;
}

:global(.datepicker-component .m-calendar tbody td:hover){
    background: #5ccdde;
    border: none;
}

:global(.datepicker-component .m-calendar .toolbar button){
    border: 1px solid #5ccdde;
    background: #5ccdde;
}

:global(.datepicker-component .m-calendar thead td){
    color: #5ccdde;
}

:global(.datepicker-component .m-calendar .toolbar .current-date){
    color: #5ccdde;
}

:global(.datepicker-component .m-input-moment .btn-save){
    display: none;
}

:global(.datepicker-component .m-time .time){
    background: #5ccdde;
}

:global(.datepicker-component .m-time .separater){
    color: #5ccdde;
}

:global(.datepicker-component .u-slider-time .value){
    background: #5ccdde;
}

:global(.datepicker-component .u-slider-time .handle:after){
    border: 3px solid #5ccdde;
    top: -13px;
}

:global(.datepicker-component .m-time .time-text){
    color: #5ccdde;
}

:global(.datepicker-component .m-calendar .current-day){
    color: #5ccdde;
}

:global(.datepicker-component .m-calendar .current-day):hover{
    color: #fff;
}

.description{
    display: block;
    color: var(--italicDescription);
    font-size: .78571428571rem;
    margin-top: 10px;
}

.success{
    box-shadow: inset 0px 0px 0px 2px #00cc44;
}

.error{
    box-shadow: inset 0px 0px 0px 2px #ec3232;
}

.descError{
    color: #ec3232;
}

.descSuccess{
    color: #00cc44;
}

@media (--xSmall){
    .datePicker{
        flex-direction: column;
    }

    .label{
        flex-basis: 20px;
    }
}

@media (--deviceXsmall){
    .datePicker{
        flex-direction: column;
    }

    .label{
        margin-left: 3px;
    }

    .calendar{
        position: fixed;
        top: 15%;
        left: 4%;
        z-index: 99;
    }

    :global(.datepicker-component .m-calendar .current-day:hover){
        color: #fff;
        background: #5ccdde;
    }

    /* :global(.m-input-moment){
        width: 90vw;
        height: 118vw;
    }

    :global(.options){
        display: flex;
        height: 12vw;
    }

    :global(.tabs){
        height: 100vw;
    }

    :global(.im-btn){
        height: 8vw;
        font-size: 3vw;                
    }

    :global(.m-calendar .toolbar){
        height: 10vw;
        line-height: 10vw;
    }

    :global(div.m-calendar td){
        height: 13vw;
        font-size: 4.5vw; 
    }
    
    :global(.current-date){
        font-size: 7vw;
        padding-top: 2vw;
    }

    :global(.m-calendar .toolbar button){
        height: 10vw;
        width: 10vw;
        font-size: 6vw;
    }

    :global(.m-input-moment .options button){
        height: 11vw;
        font-size: 5vw;
    }

    :global(.m-calendar .current-day){
        color: #5ccdde;
    }

    :global(.m-input-moment .btn-save){
        font-size: 4vw;
    }

    :global(.m-time .time){
        height: 25vw;
        width: 25vw;
        font-size: 20vw;
        line-height: normal;
    }

    :global(.m-time .separater){
        color: #5ccdde;
        font-size: 15vw;
        width: 10vw;
    }

    :global(.u-slider-time .handle:after){
        border: 3px solid #5ccdde;
        height: 5vw;
        width: 5vw;
        top: -1.8vw;
        left: -3vw;
    }

    :global(.m-time .time-text){
        font-size: 4vw;
        margin-top: 1vw;
        margin-bottom: 2vw;
    }

    :global(.u-slider-time){
        height: 2vw;
    }

    :global(.m-time .sliders){
        padding: 0px 20px;
    } */

    .cover{
        background-color: rgba(0, 0, 0, 0.7);
        animation: opa 0.2s;
    }

    @keyframes opa {
        from {opacity: 0}
        to {opacity: 1}
    }
}

@media (--deviceSmall){
    .datePicker{
        flex-direction: column;
    }

    .label{
        margin-left: 3px;
    }

    .calendar{
        position: fixed;
        top: 15%;
        left: 4%;
        z-index: 99;
    }

    :global(.datepicker-component .m-calendar .current-day:hover){
        color: #fff;
        background: #5ccdde;
    }

    /* :global(.m-input-moment){
        width: 90vw;
        height: 118vw;
    }

    :global(.options){
        display: flex;
        height: 12vw;
    }

    :global(.tabs){
        height: 100vw;
    }

    :global(.im-btn){
        height: 8vw;
        font-size: 3vw;                
    }

    :global(.m-calendar .toolbar){
        height: 10vw;
        line-height: 10vw;
    }

    :global(div.m-calendar td){
        height: 13vw;
        font-size: 4.5vw; 
    }
    
    :global(.current-date){
        font-size: 7vw;
        padding-top: 2vw;
    }

    :global(.m-calendar .toolbar button){
        height: 10vw;
        width: 10vw;
        font-size: 6vw;
    }

    :global(.m-input-moment .options button){
        height: 11vw;
        font-size: 5vw;
    }

    :global(.m-calendar .current-day){
        color: #5ccdde;
    }

    :global(.m-input-moment .btn-save){
        font-size: 4vw;
    }

    :global(.m-time .time){
        height: 25vw;
        width: 25vw;
        font-size: 20vw;
        line-height: normal;
    }

    :global(.m-time .separater){
        color: #5ccdde;
        font-size: 15vw;
        width: 10vw;
    }

    :global(.u-slider-time .handle:after){
        border: 3px solid #5ccdde;
        height: 5vw;
        width: 5vw;
        top: -1.8vw;
        left: -3vw;
    }

    :global(.m-time .time-text){
        font-size: 4vw;
        margin-top: 1vw;
        margin-bottom: 2vw;
    }

    :global(.u-slider-time){
        height: 2vw;
    }

    :global(.m-time .sliders){
        padding: 0px 20px;
    } */

    .cover{
        background-color: rgba(0, 0, 0, 0.7);
        animation: opa 0.2s;
    }

    @keyframes opa {
        from {opacity: 0}
        to {opacity: 1}
    }
}