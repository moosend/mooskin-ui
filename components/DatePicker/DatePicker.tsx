import * as React from 'react';

// Date-FNS
import DateFnsUtils from '@date-io/date-fns';

// Models
import { IDatePickerComponentProps, IDatePickerKeyboardComponentProps, PickerType } from './model';

// Material-UI Date Picker
import { DatePicker as DatePickerUI, KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';

// Components
import { Input } from '../Input/Input';
import { withMooskinContext } from '../index';
import { createMuiTheme } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/styles';
import variables from '../_utils/globals/variables';

const ComponentByType = {
	date: DatePickerUI,
	'date-keyboard': KeyboardDatePicker,
};

/**
 * DatePicker
 */
export const DatePicker: React.FC<IDatePickerComponentProps | IDatePickerKeyboardComponentProps> = withMooskinContext((props) => {
	const materialTheme = createMuiTheme({
		overrides: {
			MuiButton: {
				textPrimary: {
					backgroundColor: (props as any).palette.backgroundColors.background || variables.backgroundColors.background,
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
				label: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiPaper: {
				root: {
					backgroundColor: (props as any).palette.backgroundColors.background || variables.backgroundColors.background,
				},
			},
			MuiPickersCalendarHeader: {
				switchHeader: {
					backgroundColor: (props as any).palette.backgroundColors.background || variables.backgroundColors.background,
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
				iconButton: {
					backgroundColor:
						(props as any).palette.backgroundColors.primary1 + '!important' || variables.backgroundColors.primary1 + '!important',
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
				dayLabel: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiPickersDay: {
				current: {
					color: (props as any).palette.fontColors.white || variables.fontColors.white,
				},
				day: {
					backgroundColor: (props as any).palette.backgroundColors.background || variables.backgroundColors.background,
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
				dayDisabled: {
					color: (props as any).palette.backgroundColors.medgray1 || variables.backgroundColors.medgray1,
				},
				daySelected: {
					backgroundColor: (props as any).palette.backgroundColors.primary1 || variables.backgroundColors.primary1,
					color: (props as any).palette.fontColors.white || variables.fontColors.white,
				},
			},
			MuiTypography: {
				body2: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
				alignCenter: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiIconButton: {
				label: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiPickersModal: {
				dialogAction: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiPickersSlideTransition: {
				transitionContainer: {
					color: (props as any).palette.fontColors.text || variables.fontColors.text,
				},
			},
			MuiPickersToolbar: {
				toolbar: {
					backgroundColor: (props as any).palette.backgroundColors.primary1 || variables.backgroundColors.primary1,
				},
			},
			MuiPickersToolbarButton: {
				toolbarBtn: {
					backgroundColor: (props as any).palette.backgroundColors.primary1 || variables.backgroundColors.primary1,
				},
			},
		},
	} as any);
	const { inputProps } = props;

	const renderInput = (dateInputProps: any) => {
		return (
			<div ref={dateInputProps.inputRef}>
				<Input style={{ width: 500 }} {...dateInputProps} {...inputProps} />
			</div>
		);
	};

	const type: PickerType = props.pickerType ? props.pickerType : 'date';

	const PickerComponent = ComponentByType[type];

	return (
		<MuiPickersUtilsProvider utils={DateFnsUtils}>
			<ThemeProvider theme={materialTheme}>
				<PickerComponent {...props} variant="inline" TextFieldComponent={renderInput} />
			</ThemeProvider>
		</MuiPickersUtilsProvider>
	);
});

DatePicker.defaultProps = {
	format: 'dd/MM/yyyy',
	pickerType: 'date',
	// variant: 'inline',
};

DatePicker.displayName = 'DatePicker';
