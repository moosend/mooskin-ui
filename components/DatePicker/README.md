# MooSkin UI - DatePicker Component

The MooSkin DatePicker Component works similar to the normal HTML `<input type="date"/>` element, but with pre-defined styling and attributes.

___

### Usage

To start using the DatePicker Component first you have to Import it

```
Import {DatePicker} from 'mooskin';
```
or modular import
```
// JS
import DatePicker from 'mooskin/lib/DatePicker';

// CSS
import 'mooskin/lib/DatePicker/style.css';
```


And then you can simply start using it by typing

```
<DatePicker />
```

Like the `<input type="date"/>` element it will accept given attributes and render differently based on the given attributes

### Examples

Passing the date attribute, which requires [DayJS](https://github.com/iamkun/dayjs).

```
import dayjs from 'dayjs';

<DatePicker date={dayjs()} />
```

or making it disabled with a label, in this case the proper disabled class will be loaded (ex. disabling cursor)

```
<DatePicker label="Start date" disabled />
```

or pass it a function for event handling

```
<DatePicker onChange={yourFunc} />
```

or just give it a custom style

```
<DatePicker style={yourStyle} />
```

### Callback

The DatePicker Component Callback will always return a [DayJS](https://github.com/iamkun/dayjs) object on each change. This can be used with a function passed via the `onChange` prop. For example, if u want to console log the callback value, pass a function to the `onChange` prop.

```
const logValue = (data) => {     // data is the callback object, which consists of value and a dataLabel(not required)
    console.log(data.value._d);
};

<DatePicker onChange={logValue} />
```
In this case on each input change, the date string will be returned.

This can be used in various situations and combinations, for an enhanced development experience.

### Validation

The Datepicker Component can be validated using the `status` and `validate` props. The `status` prop will accept an 'error' or 'success' string to apply the appropriate classes to the datepicker, or none for that matter. The `validate` prop is the validation function which accepts a arg of an object containing `value`, `dataLabel` and `required` props of the component as properties, and depending on the received value, you can use any kind of validation you like. The `description` prop can also be used in the process to inform the user about validation errors. The validation will fire the first time onBlur, and every other time onChange.

#### Example

```
this.state = {
    status: '',
    message: '',
    date: ''
}

<DatePicker
    file={this.state.file}
    validate={this.validate}
    status={this.state.status}
    description={this.state.message}
    onChange={this.onChange}
/>

onChange(e, data) {
    this.setState({date: data.value})
}

validate(data){
    if (data.value < dayjs()>){
        this.setState({status: 'error', message: 'You selected a past Date'});
    } else {
        this.setState({status: '', message: ''})
    }
}
```

### Alternate selection of time

The `DateSelect` Component provides a different way of selecting minutes/hours/days/weekdays/month days from a select field based on the passed down props. The following example shows the selection of hours depending on the time format (24 hours/12 hours)

#### Example

```
<DateSelect onChange={this.onChange} format="24-Hour" value={17} type="hour" label="Hour (24H)" />

onChange(data){
    console.log(data.value);
}
```

This will render a Mooskin Select component with 24 options, one for each hour. It is in 24-Hours format so the first Option will be of the value 00, the last one will be 23.

```
<DateSelect onChange={this.onChange} format="12-Hour" type="hour" label="Hour (12H)" />

onChange(data){
    console.log(data.value);
}
```

This renders a Select component with 24 options, the twelve first options will be of the AM time, starting from 00 to 11, followed by PM time of the same options.

The days of the month Select will render as in the following example:

```
<DateSelect onChange={this.onChange} format="1" type="day" label="Day" />

onChange(data){
    console.log(data.value);
}
```

Now in this case, the component will render a Select with 31 Options, where `format` reflects the day of the month, in this case `1` for January, hence 31 options. If in the case of passing down the `format` option with `2` (February), the Select will render with 28 options, the first being `1st` with a value of 1, and the last being the `28th` with the value of 28.

#### Callback

The callback works the same as with the DatePicker component, where it returns a `data` object consisting of `value` & `dataLabel` of the component. The value will always be returned as a number. For example: on the 24-Hours format of the hours component type, when selecting `01`, will return 1 as a callback value, selecting `22` will return 22. On the 12-Hours format, selecting `02 AM` will return 2, selecting `01 PM` will return 13 and so on. On the weekly type, selecting 'Tuesday' will return 2, on the `ordinal` type, selecting `First` will return 1, selecting `Last` will return 01.

<div class="playground-doc">

## Supported attributes 

* `id` - id of the element
* `date` - selected date (DayJS object)
* `format` - format the date differently. See [format](https://github.com/iamkun/dayjs/blob/dev/docs/en/API-reference.md#format-formatstringwithtokens-string) types
* `allowInput` - allows date input through input typing, this will however ignore the format prop.
* `label` - datepicker label
* `dateOnly` - datepicker with date only
* `disabled` - input field should be disabled
* `required` - input field is required
* `className` - css class
* `dataLabel` - label what kind of data
* `nowButton` - adds a button that selects current date when clicked
* `style` - input field style
* `onChange` - callback to be triggered on input change

## DateSelect attributes

* `id` - id of the element
* `format` - formats the select options, example `12-Hour` or `24-Hour` for `hour` type. Or `1...12` for rendering day options for a month in `month` type
* `type` - 'hour' | 'minute' | 'month' | 'week' | 'ordinal'
* `label` - datepicker label
* `value` - selected value
* `labelTop` - positions the label on the top
* `className` - css class
* `dataLabel` - label what kind of data 
* `style` - input field style
* `onChange` - callback to be triggered on input change

</div>

Allthough these attributes are supported, all of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)