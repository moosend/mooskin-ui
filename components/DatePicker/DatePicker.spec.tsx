import moment from 'moment';
import * as React from 'react';
import DatePicker from './DatePicker';

import { mount, shallow } from 'enzyme';

describe('DatePicker', () => {

    test('renders correctly', () => {
        const func = jest.fn();
        Date.now = jest.fn(() => new Date(Date.UTC(2017, 0, 1)).valueOf());

        const tree = shallow(
            <DatePicker
                onChange={func}
                disabled
                className="myClass"
                style={{color: 'blue'}}
                id={'input1'}
                date={moment()}
                label="with a label ofc"
                required
            />
        );
        expect(tree).toMatchSnapshot();
    });

    test('renders properly into dom with props', () => {
        Date.now = jest.fn(() => new Date(Date.UTC(2017, 0, 1)).valueOf());

        const component = shallow(<DatePicker date={moment()} required/>);

        expect(component.find('InputMoment').prop('moment')).toEqual(moment());
        expect(component.find('input').prop('required')).toBeTruthy;
    });

    test('onChange callback is called when date is changed', () => {
        const func = jest.fn();

        const component = mount(<DatePicker onChange={func}/>);

        component.find('InputMoment').find('td').at(15).simulate('click');
        expect(func).toHaveBeenCalled();
    });

    test('onChange callback is NOT called when date is changed when disabled is passed', () => {
        const func = jest.fn();

        const component = mount(<DatePicker onChange={func} disabled/>);

        component.find('InputMoment').find('td').at(15).simulate('click');
        expect(func).toHaveBeenCalled();
    });

    test('datepicker shows when input is clicked and disappears when anywhere except datepicker is clicked', () => {
        const func = jest.fn();

        const component = shallow(<DatePicker onChange={func}/>);

        component.find('input').simulate('click');
        expect(component.find('div').at(2).prop('style')).toEqual({display: 'block'});

        component.find('div').last().simulate('click');
        expect(component.find('div').at(2).prop('style')).toEqual({display: 'none'});
    });

});
