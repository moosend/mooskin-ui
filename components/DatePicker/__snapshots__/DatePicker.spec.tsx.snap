// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker renders correctly 1`] = `
<div
  className="datepicker-component myClass datePicker"
  id="input1"
  style={
    Object {
      "color": "blue",
    }
  }
>
  <label
    className="label"
    style={Object {}}
  >
    with a label ofc
  </label>
  <div
    className="wrapper"
  >
    <input
      className="dateInput disabled"
      disabled={true}
      onClick={[Function]}
      readOnly={true}
      required={true}
      value="01 Jan 2017, 1:00"
    />
    <div
      className="calendar"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <InputMoment
        hourStep={1}
        minStep={1}
        moment={"2017-01-01T00:00:00.000Z"}
        nextMonthIcon="ion-ios-arrow-right"
        onChange={[Function]}
        onSave={[Function]}
        prevMonthIcon="ion-ios-arrow-left"
      />
      <div
        className="cover"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
